### PR TITLE
Implement SA1600, SA1601, SA1602

### LINKED ISSUE #120: SA1600: ElementsMustBeDocumented
http://www.stylecop.com/docs/SA1600.html

### PR DESCRIPTION
**Pull request by pdelvo**
Fixes #120, #121, #122.

I don't have implemented code fixes because automaticly generated documentation is as good as no documentation at all

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs [hunk 1]
```diff
@@ -44,7 +50,195 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
-            // TODO: Implement analysis
+            context.RegisterSyntaxNodeAction(HandleTypeDeclaration, SyntaxKind.ClassDeclaration);
+            context.RegisterSyntaxNodeAction(HandleTypeDeclaration, SyntaxKind.StructDeclaration);
+            context.RegisterSyntaxNodeAction(HandleTypeDeclaration, SyntaxKind.InterfaceDeclaration);
+            context.RegisterSyntaxNodeAction(HandleTypeDeclaration, SyntaxKind.EnumDeclaration);
+            context.RegisterSyntaxNodeAction(HandleMethodDeclaration, SyntaxKind.MethodDeclaration);
+            context.RegisterSyntaxNodeAction(HandleConstructorDeclaration, SyntaxKind.ConstructorDeclaration);
+            context.RegisterSyntaxNodeAction(HandlePropertyDeclaration, SyntaxKind.PropertyDeclaration);
+            context.RegisterSyntaxNodeAction(HandleIndexerDeclaration, SyntaxKind.IndexerDeclaration);
+            context.RegisterSyntaxNodeAction(HandleFieldDeclaration, SyntaxKind.FieldDeclaration);
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: Fields aren't explicitly mentioned in the StyleCop documentation, but I agree it makes sense to analyze them.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs [hunk 2]
```diff
@@ -44,7 +50,195 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+            context.RegisterSyntaxNodeAction(HandleEventFieldDeclaration, SyntaxKind.EventFieldDeclaration);
+        }
+
+        private void HandleTypeDeclaration(SyntaxNodeAnalysisContext context)
+        {
+            BaseTypeDeclarationSyntax declaration = context.Node as BaseTypeDeclarationSyntax;
+
+            bool isDefinedInClassOrStruct = IsDefinedInClassOrStructRecursive(declaration.Parent);
+
+            if (declaration != null && NeedsComment(declaration.Modifiers, isDefinedInClassOrStruct ? SyntaxKind.PrivateKeyword : SyntaxKind.PublicKeyword))
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: The default visibility for a top-level type is `internal`, not `public`.

**Comment by sharwell** [comment]
@pdelvo I'm making notes, but I might just go ahead and fix these so I can merge the PR. I'll let you know when I get done if there's anything you need to do.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs [hunk 3]
```diff
@@ -44,7 +50,195 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+            {
+                var leadingTrivia = declaration.GetLeadingTrivia();
+                var commentTrivia = leadingTrivia.FirstOrDefault(x => x.IsKind(SyntaxKind.SingleLineDocumentationCommentTrivia));
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: This is also a valid documentation comment:

```csharp
/**
 * <summary>This is a documentation comment summary.</summary>
 */
public void SomeMethod() { }
```

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs [hunk 4]
```diff
@@ -44,7 +50,195 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+            if (variableDeclaration != null && NeedsComment(declaration.Modifiers, SyntaxKind.PrivateKeyword))
+            {
+                var leadingTrivia = declaration.GetLeadingTrivia();
+                var commentTrivia = leadingTrivia.FirstOrDefault(x => x.IsKind(SyntaxKind.SingleLineDocumentationCommentTrivia));
+                if (XmlCommentHelper.IsMissingOrEmpty(commentTrivia))
+                {
+                    var locations = variableDeclaration.Variables.Select(v => v.Identifier.GetLocation()).ToArray();
+                    if (locations.Length > 0)
+                    {
+                        context.ReportDiagnostic(Diagnostic.Create(Descriptor, locations.First()));
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: I believe the warning should be reported for each field, as opposed to just the first one.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs (hunk 3 above)
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: It seems like these three lines could be combined in a new helper method in `XmlCommentHelper` at the same time they're updated to handle both documentation comment formats.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs [hunk 5]
```diff
@@ -44,7 +50,195 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: I believe the warning should be reported for each event, as opposed to just the first one.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs [hunk 6]
```diff
@@ -44,7 +50,195 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+
+        private bool NeedsComment(SyntaxTokenList modifiers, SyntaxKind defaultModifier)
+        {
+
+            return (modifiers.Any(SyntaxKind.PublicKeyword)
+                || modifiers.Any(SyntaxKind.ProtectedKeyword)
+                || modifiers.Any(SyntaxKind.InternalKeyword)
+                || defaultModifier == SyntaxKind.PublicKeyword
+                || defaultModifier == SyntaxKind.ProtectedKeyword
+                || defaultModifier == SyntaxKind.InternalKeyword)
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: Are you sure this should be reported for `internal` elements?

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs [hunk 7]
```diff
@@ -44,7 +50,195 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+                    context.ReportDiagnostic(Diagnostic.Create(Descriptor, declaration.Identifier.GetLocation()));
+                }
+            }
+        }
+
+        private void HandleMethodDeclaration(SyntaxNodeAnalysisContext context)
+        {
+            MethodDeclarationSyntax declaration = context.Node as MethodDeclarationSyntax;
+
+            if (declaration != null && NeedsComment(declaration.Modifiers, SyntaxKind.PrivateKeyword))
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: The default visibility of a method declaration within an interface is `public`.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs [hunk 8]
```diff
@@ -44,7 +50,195 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+                    context.ReportDiagnostic(Diagnostic.Create(Descriptor, declaration.Identifier.GetLocation()));
+                }
+            }
+        }
+
+        private void HandlePropertyDeclaration(SyntaxNodeAnalysisContext context)
+        {
+            PropertyDeclarationSyntax declaration = context.Node as PropertyDeclarationSyntax;
+
+            if (declaration != null && NeedsComment(declaration.Modifiers, SyntaxKind.PrivateKeyword))
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: The default visibility of a property declaration within an interface is `public`.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs [hunk 9]
```diff
@@ -44,7 +50,195 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+            }
+        }
+
+
+
+        private void HandleIndexerDeclaration(SyntaxNodeAnalysisContext context)
+        {
+            IndexerDeclarationSyntax declaration = context.Node as IndexerDeclarationSyntax;
+
+            if (declaration != null && NeedsComment(declaration.Modifiers, SyntaxKind.PrivateKeyword))
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: The default visibility of an indexer declaration within an interface is `public`.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs [hunk 10]
```diff
@@ -44,7 +50,195 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+            }
+        }
+
+        private void HandleDelegateDeclaration(SyntaxNodeAnalysisContext context)
+        {
+            DelegateDeclarationSyntax declaration = context.Node as DelegateDeclarationSyntax;
+
+            bool isDefinedInClassOrStruct = IsDefinedInClassOrStructRecursive(declaration.Parent);
+
+            if (declaration != null && NeedsComment(declaration.Modifiers, isDefinedInClassOrStruct ? SyntaxKind.PrivateKeyword : SyntaxKind.PublicKeyword))
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: The default visibility of a top-level delegate declaration is `internal`.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs [hunk 11]
```diff
@@ -44,7 +50,195 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+                    context.ReportDiagnostic(Diagnostic.Create(Descriptor, declaration.Identifier.GetLocation()));
+                }
+            }
+        }
+
+        private void HandleEventDeclaration(SyntaxNodeAnalysisContext context)
+        {
+            EventDeclarationSyntax declaration = context.Node as EventDeclarationSyntax;
+
+            if (declaration != null && NeedsComment(declaration.Modifiers, SyntaxKind.PrivateKeyword))
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: The default visibility of an event declaration within an interface is `public`.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs [hunk 12]
```diff
@@ -44,7 +50,195 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+                }
+            }
+        }
+
+        private void HandleEventFieldDeclaration(SyntaxNodeAnalysisContext context)
+        {
+            EventFieldDeclarationSyntax declaration = context.Node as EventFieldDeclarationSyntax;
+            var variableDeclaration = declaration?.Declaration;
+
+            if (variableDeclaration != null && NeedsComment(declaration.Modifiers, SyntaxKind.PrivateKeyword))
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: The default visibility of an event declaration within an interface is `public`.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs [hunk 13]
```diff
@@ -17,17 +22,18 @@
     /// header is empty. In C# the following types of elements can have documentation headers: classes, constructors,
     /// delegates, enums, events, finalizers, indexers, interfaces, methods, properties, and structs.</para>
     /// </remarks>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
     public class SA1600ElementsMustBeDocumented : DiagnosticAnalyzer
     {
         public const string DiagnosticId = "SA1600";
         internal const string Title = "Elements must be documented";
-        internal const string MessageFormat = "TODO: Message format";
+        internal const string MessageFormat = "Elements must must have a non empty documentation";
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: For now I would change this to just the following:

> Elements must be documented

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs [hunk 14]
```diff
@@ -44,7 +50,195 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+            context.RegisterSyntaxNodeAction(HandleDelegateDeclaration, SyntaxKind.DelegateDeclaration);
+            context.RegisterSyntaxNodeAction(HandleEventDeclaration, SyntaxKind.EventDeclaration);
+            context.RegisterSyntaxNodeAction(HandleEventFieldDeclaration, SyntaxKind.EventFieldDeclaration);
+        }
+
+        private void HandleTypeDeclaration(SyntaxNodeAnalysisContext context)
+        {
+            BaseTypeDeclarationSyntax declaration = context.Node as BaseTypeDeclarationSyntax;
+
+            bool isDefinedInClassOrStruct = IsDefinedInClassOrStructRecursive(declaration.Parent);
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: I would refactor the word `Defined` to be `Nested`, which more closely matches the terminology in the language specification.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1601PartialElementsMustBeDocumented.cs [hunk 15]
```diff
@@ -63,7 +68,7 @@ public class SA1601PartialElementsMustBeDocumented : DiagnosticAnalyzer
     {
         public const string DiagnosticId = "SA1601";
         internal const string Title = "Partial elements must be documented";
-        internal const string MessageFormat = "TODO: Message format";
+        internal const string MessageFormat = "Partial elements must have a non empty documentation";
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1601PartialElementsMustBeDocumented.cs]: For now I would change this to just the following:

> Partial elements must be documented

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1601PartialElementsMustBeDocumented.cs [hunk 16]
```diff
@@ -86,7 +91,42 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
         /// <inheritdoc/>
         public override void Initialize(AnalysisContext context)
         {
-            // TODO: Implement analysis
+            context.RegisterSyntaxNodeAction(HandleClassDeclaration, SyntaxKind.ClassDeclaration);
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1601PartialElementsMustBeDocumented.cs]: You can also have a `partial struct` or `partial interface`.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1602EnumerationItemsMustBeDocumented.cs [hunk 17]
```diff
@@ -34,7 +39,7 @@ public class SA1602EnumerationItemsMustBeDocumented : DiagnosticAnalyzer
     {
         public const string DiagnosticId = "SA1602";
         internal const string Title = "Enumeration items must be documented";
-        internal const string MessageFormat = "TODO: Message format";
+        internal const string MessageFormat = "Enumeration items must have a non empty documentation";
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1602EnumerationItemsMustBeDocumented.cs]: For now I would change this to just the following:

> Enumeration items must be documented

**Comment by sharwell** [comment]
@pdelvo There are a number of issues I pointed out. I'll reassign this back to you to address them. Let me know if you want me to do anything or if you have any questions. :smile:

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs (hunk 6 above)
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: If we follow StyleCops implementation then it should be reported for internal members

**Comment by pdelvo** [comment]
I fixed all your remarks

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs (hunk 6 above)
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: :+1:

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs (hunk 1 above)
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: I just checked StyleCop and it reports a SA1600 for fields too.
```warning : SA1600 : CSharp.Documentation : The field must have a documentation header.```

**Comment by sharwell** [comment]
@pdelvo I sent a pull request to **your fork** so you can review my proposed changes to this (and hopefully merge them :+1: )

**Comment by oatkins** [comment]
SC has a setting that relaxes these rules to apply to public types and members only. It's "documented" in [this bug report][1]. It's a really useful feature, because it helps ensure that the public API of a library is properly documented.

I'm not sure how this project is handling external settings files, but it would be really good to have something that maintains this feature.

[1]: http://stylecop.codeplex.com/workitem/6772

**Comment by sharwell** [comment]
@oatkins I suggest you file that request as a separate issue. It's likely to get lost here.

**Comment by oatkins** [comment]
Thanks @sharwell - I added issue #789.

**Pull request merged**